File: lib/data/sync/body_energy_chain_sync_service.dart
import 'package:flutter/foundation.dart';

import '../../core/period/time_range.dart';
import '../../core/time/local_date.dart';
import '../../domain/health/health_permissions.dart';
import '../../domain/insights/body_energy_timeline.dart';
import '../../domain/model/refresh_mode.dart';
import '../prefs/preferences_repository.dart';
import '../repository/body_energy_baseline_cache_store.dart';
import '../repository/body_energy_timeline_store.dart';
import '../repository/contract/body_energy_repository.dart';
import '../repository/impl/health_connect_gating.dart';
import '../source/health/health_data_source.dart';

/// How many days back the warm window reaches. Matches the repository's chain
/// lookback, so any day inside the window finds a stored anchor.
const int bodyEnergyChainWarmDays = 14;

/// Keeps a rolling window of recent Body Energy days computed and stored, so
/// the chain the detail screen walks is almost never cold.
///
/// Body Energy carries across midnight: each day opens where the previous one
/// closed. Without a warm window, a user who last opened the app a week ago
/// would find no stored predecessor and the day would have to restart at the
/// neutral score — the foreground gap fill is deliberately bounded to two days
/// because each day costs ~8 Health Connect reads, and closing a week of them
/// while someone waits on a screen is not acceptable. That work belongs here.
///
/// Best-effort throughout, like `CaloriesHistorySyncService`: every failure is
/// swallowed and retried on the next pass, never surfaced.
class BodyEnergyChainSyncService {
  BodyEnergyChainSyncService(
    this._repository,
    this._store,
    this._baselineStore,
    this._dataSource,
    this._preferences, {
    this._clock = DateTime.now,
    this._windowDays = bodyEnergyChainWarmDays,
  });

  final BodyEnergyRepository _repository;
  final BodyEnergyTimelineStore _store;
  final BodyEnergyBaselineCacheStore _baselineStore;
  final HealthDataSource _dataSource;
  final PreferencesRepository _preferences;
  final DateTime Function() _clock;
  final int _windowDays;

  /// A cold install has to walk the whole window at ~8 Health Connect reads a
  /// day. That is fine in the background but must not run away, so a pass stops
  /// when the budget is spent and resumes where it left off — the days it
  /// already wrote are skipped as fresh next time.
  static const Duration _passBudget = Duration(seconds: 90);

  /// Every screen open calls [syncAll]; without this, opening Body Energy five
  /// times in a minute would re-walk the window five times.
  static const Duration _throttle = Duration(minutes: 30);

  /// How long a stored past day counts as fresh, matching the repository's own
  /// past-day staleness rule. Only used to skip work the repository would
  /// otherwise skip anyway; the repository stays the authority.
  static const Duration _dayFreshness = Duration(hours: 24);

  Future<void>? _running;

  /// Warm the chain. Concurrent calls share one run.
  ///
  /// Pass [force] to bypass the throttle — for a caller that has just made the
  /// stored chain wrong (a watch sync back-filling days) rather than one merely
  /// opening a screen. A forced call still joins an in-flight run: whatever is
  /// already walking will pick up the holes.
  Future<void> syncAll({bool force = false}) =>
      _running ??= _sync(force: force).whenComplete(() => _running = null);

  Future<void> _sync({bool force = false}) async {
    try {
      // One-shot cleanup of the retired SharedPreferences timelines. This is
      // the natural home for it: already best-effort, already on the UI
      // isolate, and it runs before any chain work needs the prefs.
      await _baselineStore.purgeLegacyTimelineEntries();

      final granted = await _dataSource.grantedIfAvailable();
      if (!granted.contains(HcPermissions.readHeartRate)) return;

      // Rows computed under a retired calibration are wrong, not merely stale,
      // so a signature change purges rather than letting them age out.
      final signature = _globalSignature(granted);
      if (await _store.storedGlobalSignature() != signature) {
        await _store.purgeAll();
        await _store.writeGlobalSignature(signature);
      }

      final now = _clock();
      final lastPass = await _store.lastPassAt();
      if (!force && lastPass != null && now.difference(lastPass) < _throttle) {
        return;
      }

      final today = LocalDate.fromDateTime(now);
      await _store.applyRetention(today);

      // Days the repository would serve from storage cost nothing to keep.
      // Skipping them here rather than letting its cache check do it saves a
      // permission round-trip and a store read per day — on the common warm
      // pass that is the difference between a dozen calls and none.
      //
      // The rule must match the repository's, or this would keep recomputing
      // settled days it would happily have served: a day is worth revisiting
      // only while it can still gain late-arriving data, or while today's copy
      // of it has aged past a day.
      final window = await _store.storedDaysBetween(
        today.minusDays(_windowDays - 1),
        today.minusDays(1),
      );
      final freshEpochDays = {
        for (final day in window)
          if (today.epochDay - day.date.epochDay > bodyEnergyChainSettlingDays ||
              now.difference(day.generatedAt) < _dayFreshness)
            day.date.epochDay,
      };

      // Oldest first, and that order is load-bearing: each day's seed must
      // already be stored by the time its successor is computed.
      final stopwatch = Stopwatch()..start();
      var completed = true;
      for (var back = _windowDays - 1; back >= 1; back--) {
        if (stopwatch.elapsed >= _passBudget) {
          completed = false;
          break;
        }
        // Today is skipped: the foreground load owns it, and recomputing it
        // here would fight its 15-minute freshness window.
        final date = today.minusDays(back);
        if (freshEpochDays.contains(date.epochDay)) continue;
        await _repository.loadTimeline(
          BodyEnergyTimelineQuery(
            period: DatePeriod(date, date),
            range: TimeRange.day,
            refreshMode: RefreshMode.normal,
          ),
        );
      }

      // Only a completed pass resets the throttle; a budget-truncated one lets
      // the next open pick up the remaining days immediately.
      if (completed) await _store.writeLastPassAt(now);
    } catch (e, s) {
      debugPrint(
        'BodyEnergyChainSyncService: warm pass failed, will retry: $e\n$s',
      );
    }
  }

  /// The chain-wide validity stamp: algorithm version plus the configured and
  /// permission inputs every day shares. The per-day signature additionally
  /// folds in the body profile, whose value varies by date — that belongs on
  /// the row, not here.
  ///
  /// The learned gains are deliberately NOT in it, because a mismatch here
  /// purges every stored day and every stored bucket. The watch fit nudges a
  /// gain by a fraction of a percent whenever it absorbs an observation, so
  /// including them meant the entire history — up to
  /// [bodyEnergyBucketRetentionDays] of it — was deleted on essentially every
  /// watch sync, which is no way to build the weekly view those buckets exist
  /// for. A gain change does not make a stored row wrong enough to destroy it:
  /// the per-day signature still refuses to SERVE one, so it is recomputed on
  /// demand, and the seed lookup can still anchor on it in the meantime.
  String _globalSignature(Set<String> granted) {
    final permissions = granted.toList()..sort();
    final configured =
        _preferences.bodyEnergyCalibrationListenable.value.zoneSignature();
    return 'v$bodyEnergyTimelineAlgorithmVersion'
        '|${configured.hashCode}|${permissions.join(',').hashCode}';
  }
}

File: app/src/main/kotlin/tech/mmarca/openvitals/data/sync/BodyEnergyChainSyncService.kt
package tech.mmarca.openvitals.data.sync

import android.util.Log
import androidx.health.connect.client.permission.HealthPermission
import androidx.health.connect.client.records.HeartRateRecord
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import tech.mmarca.openvitals.core.period.DatePeriod
import tech.mmarca.openvitals.core.period.TimeRange
import tech.mmarca.openvitals.data.repository.BodyEnergyBaselineCacheStore
import tech.mmarca.openvitals.data.repository.BodyEnergyChainSettlingDays
import tech.mmarca.openvitals.data.repository.BodyEnergyTimelineStore
import tech.mmarca.openvitals.data.repository.PreferencesRepository
import tech.mmarca.openvitals.data.repository.contract.BodyEnergyRepository
import tech.mmarca.openvitals.data.repository.contract.BodyEnergyTimelineQuery
import tech.mmarca.openvitals.data.repository.contract.HealthRepository
import tech.mmarca.openvitals.domain.insights.BodyEnergyTimelineAlgorithmVersion
import tech.mmarca.openvitals.domain.model.HealthConnectAvailability
import tech.mmarca.openvitals.domain.model.RefreshMode

/** How many days back the warm window reaches. Matches the repository's chain lookback. */
const val BodyEnergyChainWarmDays = 14L

/**
 * Keeps a rolling window of Body Energy days computed and stored, so the
 * chain is almost never cold. The foreground fill is bounded to two days;
 * closing a week belongs here. Best-effort: failures are retried next pass.
 */
@Singleton
class BodyEnergyChainSyncService(
    private val repository: BodyEnergyRepository,
    private val store: BodyEnergyTimelineStore,
    private val baselineStore: BodyEnergyBaselineCacheStore,
    private val healthRepository: HealthRepository,
    private val preferencesRepository: PreferencesRepository,
    private val clock: () -> Instant = Instant::now,
    private val zone: ZoneId = ZoneId.systemDefault(),
    private val windowDays: Long = BodyEnergyChainWarmDays,
    private val elapsedMillis: () -> Long = { System.nanoTime() / 1_000_000L },
) {
    @Inject
    constructor(
        repository: BodyEnergyRepository,
        store: BodyEnergyTimelineStore,
        baselineStore: BodyEnergyBaselineCacheStore,
        healthRepository: HealthRepository,
        preferencesRepository: PreferencesRepository,
    ) : this(
        repository = repository,
        store = store,
        baselineStore = baselineStore,
        healthRepository = healthRepository,
        preferencesRepository = preferencesRepository,
        clock = Instant::now,
    )

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val lock = Mutex()
    private var inFlight: Deferred<Unit>? = null

    /**
     * Warms the chain. Concurrent calls share one run. [force] bypasses the
     * throttle for a caller that just made the stored chain wrong.
     */
    suspend fun syncAll(force: Boolean = false) {
        val deferred = lock.withLock {
            inFlight?.takeIf { it.isActive } ?: newRun(force).also { inFlight = it }
        }
        deferred.await()
    }

    private fun newRun(force: Boolean): Deferred<Unit> {
        lateinit var created: Deferred<Unit>
        created = scope.async(start = CoroutineStart.LAZY) {
            try {
                sync(force)
            } finally {
                lock.withLock { if (inFlight === created) inFlight = null }
            }
        }
        return created
    }

    private suspend fun sync(force: Boolean) {
        try {
            // One-shot cleanup of the retired SharedPreferences timelines.
            baselineStore.purgeLegacyTimelineEntries()

            if (healthRepository.availability() != HealthConnectAvailability.AVAILABLE) return
            val granted = healthRepository.grantedPermissions()
            if (ReadHeartRatePermission !in granted) return

            // Rows under a retired calibration are wrong, not stale: purge them.
            val signature = globalSignature(granted)
            if (store.storedGlobalSignature() != signature) {
                store.purgeAll()
                store.writeGlobalSignature(signature)
            }

            val now = clock()
            val lastPass = store.lastPassAt()
            if (!force && lastPass != null && Duration.between(lastPass, now) < Throttle) return

            val today = now.atZone(zone).toLocalDate()
            store.applyRetention(today)

            // Skip days the repository would serve from storage. The rule must
            // match the repository's: a day is worth revisiting only while it can
            // still gain late data, or once today's copy has aged past a day.
            val window = store.storedDaysBetween(
                today.minusDays(windowDays - 1),
                today.minusDays(1),
            )
            val freshEpochDays = window.filter { day ->
                today.toEpochDay() - day.date.toEpochDay() > BodyEnergyChainSettlingDays ||
                    Duration.between(day.generatedAt, now) < DayFreshness
            }.mapTo(mutableSetOf()) { it.date.toEpochDay() }

            // Oldest first: each day's seed must be stored before its successor.
            val startedAt = elapsedMillis()
            var completed = true
            for (back in (windowDays - 1) downTo 1L) {
                if (elapsedMillis() - startedAt >= PassBudgetMillis) {
                    completed = false
                    break
                }
                // Today is skipped: the foreground load owns it.
                val date = today.minusDays(back)
                if (date.toEpochDay() in freshEpochDays) continue
                repository.loadTimeline(
                    BodyEnergyTimelineQuery(
                        period = DatePeriod(date, date),
                        range = TimeRange.DAY,
                        refreshMode = RefreshMode.NORMAL,
                    )
                )
            }

            // Only a completed pass resets the throttle.
            if (completed) store.writeLastPassAt(now)
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (t: Throwable) {
            Log.w(TAG, "Body Energy warm pass failed, will retry", t)
        }
    }

    /**
     * The chain-wide validity stamp: algorithm version plus the shared
     * inputs. The learned gains are left out: a mismatch purges every stored
     * day, and the watch fit nudges gains on every sync.
     */
    private fun globalSignature(granted: Set<String>): String {
        val permissions = granted.sorted().joinToString(",")
        val configured = preferencesRepository.bodyEnergyCalibration().zoneSignature()
        return "v$BodyEnergyTimelineAlgorithmVersion|${configured.hashCode()}|${permissions.hashCode()}"
    }

    private companion object {
        private const val TAG = "BodyEnergyChainSync"

        /** A cold install walks the whole window at ~8 reads a day. A pass stops at the budget and resumes. */
        const val PassBudgetMillis = 90_000L

        /** Every screen open calls `syncAll`; this keeps five opens from five walks. */
        val Throttle: Duration = Duration.ofMinutes(30)

        /** How long a stored past day counts as fresh, matching the repository. */
        val DayFreshness: Duration = Duration.ofHours(24)

        val ReadHeartRatePermission: String =
            HealthPermission.getReadPermission(HeartRateRecord::class)
    }
}
